hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerControl.java

```diff
@@ -129,73 +135,86 @@ private synchronized void process(TcpServerConnection connection, MemberHandshak
                     + " planeIndex:" + handshake.getPlaneIndex());
         }
-        boolean registered = connectionManager.register(remoteEndpoint, connection, handshake.getPlaneIndex());
 
-        if (remoteAddressAliases != null && registered) {
-            for (Address remoteAddressAlias : remoteAddressAliases) {
-                if (logger.isLoggable(Level.FINEST)) {
-                    logger.finest("Registering connection " + connection + " to address alias " + remoteAddressAlias
-                            + " planeIndex:" + handshake.getPlaneIndex());
-                }
-                connectionManager.planes[handshake.getPlaneIndex()].putConnectionIfAbsent(remoteAddressAlias, connection);
-            }
-        }
+        connectionManager.register(
+                primaryAddress,
+                connectedAddress,
+                remoteAddressAliases,
+                remoteUuid,
+                connection,
+                handshake.getPlaneIndex()
+        );
     }
 
-    private boolean checkAlreadyConnected(TcpServerConnection connection, Address remoteEndPoint, int planeIndex) {
-        Connection existingConnection = connectionManager.planes[planeIndex].getConnection(remoteEndPoint);
+    /**
+     * @param connection a new connection whose handshake message is being processed
+     * @param remoteUuid the member uuid of the remote side of the connection
+     * @param planeIndex the plane index to register connection
+     * @return true if the registration of the new incoming connection will continue
+     *          false if this newly created connection will be closed
+     */
+    private boolean handleDuplicateConnections(TcpServerConnection connection, UUID remoteUuid, int planeIndex) {
+        TcpServerConnection existingConnection = connectionManager.planes[planeIndex].getConnection(remoteUuid);
         if (existingConnection != null && existingConnection.isAlive()) {
             if (existingConnection != connection) {
-                if (logger.isFinestEnabled()) {
-                    logger.finest(existingConnection + " is already bound to " + remoteEndPoint
-                            + ", new one is " + connection + " planeIndex:" + planeIndex);
+                // If there are duplicate connections, close the connection of which the member with
+                // the smaller uuid is on the acceptor side
+                boolean isLocalMemberUuidSmallerThanRemote = serverContext.getThisUuid().compareTo(remoteUuid) < 0;
+                if (isLocalMemberUuidSmallerThanRemote == connection.isAccepted()) {
+                    connection.close(
+                            "Duplicate connection to the Member[uuid= " + remoteUuid + "]"
+                                    + " on planeIndex=" + planeIndex,
+                            null
+                    );
+                    return false;
+                } else if (isLocalMemberUuidSmallerThanRemote == existingConnection.isAccepted()) {
```

Comment:
simple `else` could be sufficient

Reply:
This duplicate handling only resolves the issue when these duplicate connections occur as a result of two members trying to establish a connection to each other at the same time. I'm thinking about a duplicate connections occur when requested address in `getOrConnect(Address)` wasn't registered before (this is only possible when using multiple addresses referencing same host machine in the network config)?